package "stabilitytest"

import "nova/network/ServerSocket"
import "nova/network/ClientSocket"
import "nova/network/ConnectionSocket"

import "novex/nest/Nest"
import "novex/nest/TestSuite"

import static "novex/nest/Nest"

testable class NetworkStability {
    visible static final String received = "THIS IS THE STRING THAT SHOULD BE RECEIVEDTHIS IS THE STRING THAT SHOULD BE RECEIVEDTHIS IS THE STRING THAT SHOULD BE RECEIVEDTHIS IS THE STRING THAT SHOULD BE RECEIVEDTHIS IS THE STRING THAT SHOULD BE RECEIVEDTHIS IS THE STRING THAT SHOULD BE RECEIVEDTHIS IS THE STRING THAT SHOULD BE RECEIVED"
    
    visible static final Int PORT = 24243
    
    [Test "Testing creating a ServerSocket"]
    createServer() -> ServerSocket {
        let socket = new ServerSocket()
        
        expectValue(socket.start(PORT), "Unable to start server on port #PORT")
        expectTrue(socket.close(), "Failed to close server socket on port #PORT")
    }
    
    [Test message: false]
    public test() {
        Console.writeLine("Testing network stability")
        
        ServerSocket server = new ServerSocket().start(PORT)
        
        ClientThread thread = new ClientThread(PORT)
        
        thread.start()
        
        Console.writeLine("Accepting ClientSocket connection")
        
        ConnectionSocket connection = server.acceptClient()
        
        expectValue(connection, "Failed to accept the client")
        
        Console.writeLine("Sending String to ClientSocket... ")
        
        expectValue(connection.out.write(received), "Server unable to send data to client")
        
        Console.writeLine("Waiting for String from ClientSocket... ")
        
        String s = connection.in.readString()
        
        expectTrue(s.count == received.count && s == received, "Server unable to receive correct message from client. Expected message of size #received.count, but received of size #s.count")
        
        thread.join()
        
        Console.write("Attempting to close ServerSocket connection... ")
        
        expectTrue(server.close(), "Unable to close server")

        Console.writeLine("Success")
    }
}